python/lsst/pipe/base/_status.py

```diff
@@ -73,6 +81,68 @@ class RepeatableQuantumError(RuntimeError):
 
     EXIT_CODE = 20
 
+    @property
+    @abc.abstractmethod
+    def metadata(self):
+        """
+        Metadata from the raising `~lsst.pipe.base.Task` with more information
+        about the failure. The contents of the dict are `~lsst.pipe.base.Task`
+        -dependent.
+        """
+        pass
+
+
+class UnprocessableDataError(RepeatableQuantumError):
+    """Exception that will be subclassed and raised by Tasks to indicate a
+    failure to process their inputs for some reason that is non-recoverable.
+
+    Notes
+    -----
+    An example is a known bright star that causes PSF measurement to fail, and
+    that makes that detector entirely non-recoverable.
+
+    Do not raise this unless we are convinced that the data cannot be
+    processed, even by a better algorithm.
+    """
+
+
+class AnnotatedPartialOutputsError(RepeatableQuantumError):
+    """Exception that runQuantum raises when the (partial) outputs it has
+    written contain information about their own incompleteness or degraded
+    quality.
+
+    This exception should always chain the original  error. When the
+    executor catches this exception, it will report the original exception. In
+    contrast, other exceptions raised from ``runQuantum`` are considered to
+    invalidate any outputs that are already written.
+    This allows a failure to be "softened" in ``run``, but still raised in
+    ``runQuantum``, to allow partial outputs to still be written.
+    """
+    @classmethod
+    def annotate(cls, exception: Exception, *,
+                 task, exposures=(), catalogs=()) -> None:
+        """Set metadata on outputs to explain the nature of the failure.
+
+        Notes
+        -----
+        This can only be called from within an except block that has caught
+        the given exception.
+        """
+        # NOTE: can't test this in pipe_base because afw is not a dependency!
```

Comment:
It seemed like there were some subtle differences between TaskMetadata and PropertyList in what the result looked like, but maybe I was just confused about the TaskMetadata nested API. I was also surprised by the fact that there wasn't a `metadata` property on Exposure/Catalog. Those both make me worried about a mock-like test in this package.

Reply:
TaskMetadata is not meant to exactly match PropertySet behavior (not PropertyList since that is not hierarchical). It exists solely so that pipe_base didn't have to depend on daf_base but people could still use similar APIs as they always had when creating the output task metadata files. I think I need to re-read the RFC because I'm confused as to why PropertyList/Set is relevant to error annotations and task exception tracking.

Reply:
`AnnotatedPartialOutputsError.annotate` annotates the task outputs and calls `task.annotate_failure` to annotate the task metadata itself. Those are `PropertyList` and `TaskMetadata`, respectively.

Reply:
I'm now really confused because middleware can not import anything from daf_base and runQuantum can't know anything about PropertyList.

Reply:
Yes, that's the point of the code comment: the new `annotate_failure` method on `task` adds fields to metadata on afw objects, if they are passed in. It's an implicit dependency, but only if someone passes those objects in and they contain a `getMetadata()` method.

Reply:
These exact questions are causing me trouble with the type annotation in this file: what type do I annotate for `exposures` and `catalogs`, given that afw is not (and should not be) a dependency here?

Reply:
I hadn't appreciated the dependency issue until @timj raised it, and I do agree it's a problem.  An implicit dependency is not nearly as bad as a direct one, but it'd be much better to formalize a dependency in the right direction by having `pipe_base` declare an interface that `afw` objects can implement, and that will solve your typing problem.  Given that the `afw` objects are C++ things that can't inherit from Python ABCs, this is a job for `typing.Protocol`.  I'm thinking something like this (probably best put into `_task_metadata.py`):

```
from collections.abc import Set
from typing import Protocol, Self

class MetadataMapping(Protocol):
     """Interface satisfied by `PropertyList`, `PropertySet`, `TaskMetadata`,
     and any nested `collections.abc.Mapping` with `str` keys and
     sufficiently simple value types.

    Objects that satisfy this interface may be assigned as values to
    `MetadataContainer` objects.
    """

    def keys(self) -> Set[str]:
        ...

    def __getitem__(self, key: str) -> int | str | float | MetadataMapping:
        ...

class MetadataContainer(MetadataGetItem, Protocol): 
    """Interface satisfied by `PropertyList`, `PropertySet`, and `TaskMetadata`.
    """

    def keys(self) -> Set[str]:
        ...

    def __getitem__(self, key: str) -> int | str | float | MetadataContainer:
        ...

    # the whole reason `MetadataMapping` exists is to capture the fact that you can pass `dict` here.
    def __setitem__(self, key: str, value: int | str | float | MetadataMapping) -> None:
       ...

    # one more method; see comment thread on PropertyList/TaskMetadata differences.
```
We could add more methods (and inheritance from another `Protocol`, `PropertySetLike`, which is defined in `_task_metadata.py`) for future use, but I think this is all this branch requires.

Once we have that, we can add another `Protocol` that `Task`, `Exposure`, and `SourceCatalog` could all satisfy:
```
class HasMetadata(Protocol):
    metadata: MetadataContainer
```
If `Exposure` and `SourceCatalog` do not have read-write `metadata` properties they should, and I consider adding them better than putting `getMetadata` and `setMetadata` into a `Protocol` in `lsst.pipe.base`.

I understand if this is rather more typing-fu that you want to include on this branch (especially since the above snippets are untested first drafts); if you want to make me another ticket I can try to get that stuff in shape for you to rebase on.

Reply:
If this means that the new interface doesn't require someone to call `PropertyList`, `getMetadata` and `setMetadata` inside `runQuantum` then that sounds great.

Reply:
@TallJimbo, done: [DM-42928](https://jira.lsstcorp.org/browse/DM-42928)

Reply:
I've put some new methods and `Protocols` up on #403 (with matching methods in [daf_base|https://github.com/lsst/daf_base/pull/82/files] and [afw](https://github.com/lsst/afw/pull/727)), but they're not quite what I wrote above.  Instead, I've just added `get_dict` and `set_dict` methods that set and get true `dict` objects with arbitrary nesting in a consistent way on all three types (`TaskMetadata`, `PropertySet`, and `PropertyList`).  The idea is that you'd package up everything you want into one nested  `dict` (exception message, type, and error metadata), and stuff all of that into the `.metadata` attribute at once.  Something like this:

```
from ._task_metadata import NestedMetadataDict

class RepeatableQuantumError(RuntimeError, abc.ABC):

    @property
    @abc.abstractmethod
    def metadata(self) -> NestedMetadataDict | None:
        """Metadata from the raising `~lsst.pipe.base.Task` with more information
        about the failure. The contents of the dict are `~lsst.pipe.base.Task`
        -dependent, and must have `str` keys and `str`, `int`, `float`, `bool`, or
        nested-dictionary (with the same key and value types) values.
        """
        raise NotImplementedError()

class AnnotatedPartialOutputsError(RepeatableQuantumError):
    ...

    @classmethod
    def annotate(cls, exception: Exception, *args: GetSetDictMetadataHolder) -> None:
        failure_info = {
            "message": str(exception),
            "type": introspection.get_full_type_name(exception),
        }
        if other := getattr(exception, "metadata", None):
            failure_info["metadata"] = other
        for item in args:
            # Some outputs may not exist, so we cannot set metadata on them; ignore
            # those here for convenience.
            if item is None or item.metadata is None:
                continue
            item.metadata.set_dict("failure", failure_info)
```

One nice thing about the new methods and properties is that they let you treat tasks, exposures, and catalogs identically, so there's no need for separate kwargs that prejudge the kinds of things we might want to annotate.